### ISSUE TITLE
Possible memory issue

### ISSUE THREAD

**Issue by egh** [thoughtbot/ember-cli-rails#451]
**Description**

While debugging some memory issues in our application, I came across a possible memory issue in ember-cli-rails. It seems that ember-cli-rails is recompiling the ember template every time a request is made, rather than reusing the cached the erb template. This is a problem both because it is slow and because rails caches the compiled template. Re-compiling every time will add to the memory of the usage of the application over time.

This can be confirmed either by adding a breakpoint to line 295 of `actionview-4.2.6/lib/action_view/template.rb` or by checking the memory usage using derailed.

This is a script to reproduce the issue in a brand new ember-cli-rails app:

```
#/bin/sh -xe
rails new ember_cli
cd ember_cli
echo 'gem "ember-cli-rails"' >> Gemfile
bundle install
ember new frontend --skip-git
rails generate ember:init
cd frontend
ember install ember-cli-rails-addon
cd ..
cat > config/database.yml <<EOF
production:
  adapter: sqlite3
  database: db/prod.sqlite3
  pool: 5
  timeout: 5000
EOF
cat > config/routes.rb<<EOF
Rails.application.routes.draw do
  mount_ember_app :frontend, to: "/"
end
EOF
cat > config/secrets.yml<<EOF
production:
  secret_key_base: 39fbddcabac94a6365f87044451c9bbb3c29b5fe8c70c27edada5ca624e5dbb9ab247f8b811ea2e20244c7d084a8c3d5d86f40ae6a198a0df28700ec4fba32a5
EOF

rake ember:install
rake ember:compile

# benchmark

echo 'gem "derailed"' >> Gemfile
bundle install
```

After generating the app, if can be tested using:
`TEST_COUNT=1000 RAILS_ENV=production bundle exec derailed exec perf:objects`

This will the root of the app 1000 times and report back on the memory usage. Of particular interest is the line:

> Total retained:  3411369 bytes (9877 objects)

and:

> retained memory by location
> -----------------------------------
>    2928842  /home/egh/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/actionview-4.2.6/lib/action_view/template.rb:296
>     240000  /home/egh/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/actionview-4.2.6/lib/action_view/template.rb:107
>     133080  /home/egh/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/actionview-4.2.6/lib/action_view/template.rb:321
>      40000  /home/egh/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/actionview-4.2.6/lib/action_view/template.rb:297

If we compare with a plain vanilla rails app:

```
#/bin/sh -xe
rails new raw_rails
cd raw_rails
bundle install
cat > config/database.yml <<EOF
production:
  adapter: sqlite3
  database: db/prod.sqlite3
  pool: 5
  timeout: 5000
EOF
cat > config/routes.rb<<EOF
Rails.application.routes.draw do
  root 'root#index'
end
EOF
cat > config/secrets.yml<<EOF
production:
  secret_key_base: 39fbddcabac94a6365f87044451c9bbb3c29b5fe8c70c27edada5ca624e5dbb9ab247f8b811ea2e20244c7d084a8c3d5d86f40ae6a198a0df28700ec4fba32a5
EOF

mkdir -p app/views/root

cat > app/views/root/index.html.erb<<EOF
foo
EOF

cat > app/controllers/root_controller.rb<<EOF
class RootController < ApplicationController
  def index
  end
end
EOF

# benchmark

echo 'gem "derailed"' >> Gemfile
bundle install
```

and run the same test:

`TEST_COUNT=1000 RAILS_ENV=production bundle exec derailed exec perf:objects`

you will see no retained memory.

It is also possible to see memory increasing by hitting the app directly:

```
for run in {1..10000} ; do curl -s 'http://localhost:3000/' -H 'Host: localhost:3000' -H 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8' -H 'Accept-Language: en-US,en;q=0.7,es;q=0.3' -H 'Connection: keep-alive' -H 'If-None-Match: W/"0da4937bb683b1714cbb6f81c6b5247f"' -H 'Cache-Control: max-age=0' > /dev/null ; done
```

Thank you for all your work on ember-cli-rails!

**Which operating system and version is the project developed on?**

Ubuntu 16.04

**Which version of [`ruby`][ruby] is the project developed on?**
ruby 2.2.4p230 (2015-12-16 revision 53155) [x86_64-linux]

**Which version of [`npm`][npm] is the project developed on?**

npm@2.14.12 /usr/local/stow/node-v4.2.4-linux-x64/lib/node_modules/npm

**Which version of [`ember-cli`][ember-cli] is the project developed on?**
version: 1.13.13
Could not find watchman, falling back to NodeWatcher for file system events.
Visit http://www.ember-cli.com/user-guide/#watchman for more info.
node: 4.2.4
npm: 2.14.10
os: linux x64

**What is the [`rails`][rails] version?**
Rails 4.2.6

**What is the [`ember-cli-rails`][gem] version (from `Gemfile`)?**
0.7.3

**What is the [`ember-cli-rails-addon`][addon] version (from `package.json`)?**
0.7.0

**Is your application server [multi-threaded][] (such as `puma` and `unicorn`) or is
it multi-process (such as thin and webrick)?**

multi-process

**What are the contents of `config/initializers/ember.rb`?**
```
EmberCli.configure do |c|
  c.app :frontend
end
```

**What are the contents of the Rails' view that renders the Ember application?**

none

**How are the EmberCLI-related routes defined?**
```
Rails.application.routes.draw do
  mount_ember_app :frontend, to: "/"
end
```
**How is the application deployed?**

not deployed

[ruby]: https://www.ruby-lang.org/
[node]: https://nodejs.org/en/
[npm]: https://www.npmjs.com/
[ember-cli]: http://ember-cli.com/
[rails]: https://github.com/rails/rails
[gem]: https://github.com/thoughtbot/ember-cli-rails
[addon]: https://github.com/rondale-sc/ember-cli-rails-addon/
[multi-threaded]: https://github.com/thoughtbot/ember-cli-rails/commit/ef0a49546a8c2e5fb0a298ba455fba97289065a1

Total retained:  3409639 bytes (9878 objects)

**Comment by egh** [comment]
It turns out that this is not a leak, despite deraileds claim that the memory is "retained". It is cleaned up by the gc. However, compiling the template does take more memory and slightly more time. See #452 for details.

**Comment by seanpdoyle** [comment]
Closing this, as it might have been resolved by #452.

@egh is this still an issue?
